### ISSUE TITLE
loadbalancing exporter[>=v0.94.0] has memory issue

### ISSUE THREAD

**Issue by alburthoffman** [open-telemetry/opentelemetry-collector-contrib#33020]
### Component(s)

exporter/loadbalancing

## Description
We tried to upgrade our otel collector in prod and started see memory increased a lot and later pods were OOMKilled.
And when rollback the pods become stable and memory reduce to 1/5.

## Steps to Reproduce
1. enable loadbalancing to export traces
2. upgrade to the version >=0.94.0
3. compare the the memory usage
4. run for a while and rollback.

## Expected Result
the memory should increase slowly comparing to old version.

## Actual Result
when rolling back, we will see
<img width="1140" alt="image" src="https://github.com/open-telemetry/opentelemetry-collector-contrib/assets/863684/ff338ddb-5cdb-4ce2-ac67-04a2c4eb0777">

### Collector version

>=0.94.0

## Environment
OS: alpine 3.12
Compiler(if manually compiled): go 1.21.0

### OpenTelemetry Collector configuration

```yaml
extensions:
  health_check:
  pprof:
    endpoint: 0.0.0.0:1777
  zpages:
    endpoint: 0.0.0.0:55679
receivers:
  jaeger:
    protocols:
      thrift_http:
        endpoint: 0.0.0.0:14268
  zipkin:
  otlp:
    protocols:
      grpc:
exporters:
  prometheus:
    endpoint: "0.0.0.0:8889"
  loadbalancing:
    routing_key: "service"
    protocol:
      otlp:
        sending_queue:
          enabled: true
          num_consumers: 300
          queue_size: 200000
        tls:
          insecure: true
        timeout: 10s
    resolver:
      dns:
        hostname: span-processor
        port: 4317

processors:
  batch:
    timeout: 5s
    send_batch_size: 800
  memory_limiter:
    check_interval: 1s
    limit_percentage: 90
    spike_limit_percentage: 5
  k8sattributes:
    passthrough: false
    extract:
      metadata:
        - k8s.namespace.name
        - k8s.pod.name
        - k8s.node.name
        - k8s.pod.hostname
    pod_association:
      - sources:
          - from: resource_attribute
            name: k8s.pod.name
service:
  pipelines:
    traces:
      receivers: [jaeger,otlp]
      processors: [memory_limiter, batch, resource, k8sattributes]
      exporters: [loadbalancing]
  extensions: [health_check, pprof, zpages]
```

**Comment by alburthoffman** [comment]
the new code which was changed at early this year tries to buffer a group of batches and then flush. this would increase the memory a lot because we are expecting one batch each time and now it merge them. This causes lots of memory usage when marshalling.

this is the allocation profile:
<img width="599" alt="image" src="https://github.com/open-telemetry/opentelemetry-collector-contrib/assets/863684/dce5b3a6-3cf6-48ed-bf70-960e1d32104a">

and this is the heap profile:
<img width="528" alt="image" src="https://github.com/open-telemetry/opentelemetry-collector-contrib/assets/863684/9042b7be-56e2-4dff-bb2f-021267b527b4">

**Comment by alburthoffman** [comment]
When changing the ConsumeTraces back to the old way without merging, problem will be solved:
<img width="892" alt="image" src="https://github.com/open-telemetry/opentelemetry-collector-contrib/assets/863684/78787abd-8916-4337-974e-c35156b66574">

and this is https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.93.0/exporter/loadbalancingexporter/trace_exporter.go#L79

**Comment by alburthoffman** [comment]
But when I tried to downgrade loadbalancingexporter to 0.93.0
<img width="1242" alt="image" src="https://github.com/open-telemetry/opentelemetry-collector-contrib/assets/863684/bbaf6388-9899-4afe-ad02-216b913c879c">

the project wont be able to compiled:
<img width="1060" alt="image" src="https://github.com/open-telemetry/opentelemetry-collector-contrib/assets/863684/8f0c9664-ce60-4394-9fcd-289f2aaed3aa">

**Comment by alburthoffman** [comment]
Could anyone take a look at this issue? I can raise a PR to fix this.

**Comment by pingping95** [comment]
i got same issue when upgrading to 0.99 and rolebacked to 0.89.

certainly memory leak occurs

**Comment by alburthoffman** [comment]
this PR fix the memory issue: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32032

**Comment by jpkrohling** [comment]
I got confirmation from two people that the linked PR above does fix the memory issue. The fix is being released as part of v0.101.0. I'm closing this issue, but if you still see the same problem with collectors at that version or newer, leave a comment and I'll reopen.
